test/e2e/pkg/infra/digitalocean/digitalocean.go

```diff
@@ -60,25 +91,58 @@ func (p Provider) writePlaybook(yaml string, starting bool) error {
 	return nil
 }
 
-// file as bytes to be written out to disk.
-// ansibleStartBytes generates an Ansible playbook to start the network
-func ansibleSystemdBytes(starting bool) string {
-	startStop := "stopped"
-	if starting {
-		startStop = "started"
-	}
-	playbook := fmt.Sprintf(`- name: start/stop testapp
+const basePlaybook = `- name: e2e custom playbook
```

Comment:
While I'm pretty confident that this would work, it is rather difficult to know what the final playbook for each provider method would look like.

Would it not make it more readable to define a [template](https://pkg.go.dev/text/template) per provider method (e.g. one for `StartNodes`, one for `StopTestnet`, one for `Reconnect`, etc.) and simply parameterize the parts that need to be variable?

Reply:
> making it more difficult to debug in case something goes wrong

Regarding this, check my comment above

Yes, I thought about using the template library, but in the end I decided to go simple. As we add more logic from `qa-infra`, we can evaluate whether to move over to templates.

Reply:
Interestingly I would consider using the template library to be the simpler option from a coding/debugging perspective. My concern's not so much with the generated code as it is with the Go code and the strings within the Go code.

For instance, some of the string concatenations are performed with strings with indentation where that indentation has meaning. It's usually far easier to reason about and extend when the whole template's visible, like with the Docker compose template: https://github.com/cometbft/cometbft/blob/dc28d6bd982b48a446ef9299cfc241d0da075c78/test/e2e/pkg/infra/docker/docker.go#L53-L117

For the current code, we need to now know that some strings require indentation with 4 spaces, others with 2 spaces, etc., making it more difficult to debug/extend since you can't see those strings within their intended context.

Reply:
You've got a point with the indentation. Current approach is not the best for that, I agree.
There's also another consideration that made me lean toward the current approach (for the time being). Playbooks have two parts:

* the initial part which is kind of static, so it renders itself well for templating
* the tasks, which are additive, so it's less obvious to me how a single template would help (unless you duplicate contents in many templates, as we have it in `qa-infra`)

Maybe, a solution that represents a best balance would be to have a template for the beginning, and then other templates, one for each of the different kind of tasks we support (shell, systemd, apt, etc).

However, I'd like to defer this to the next round of improvements.